ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,281 @@
+		instanceOwnership += tokenOwnership
+		tokensQueue.Add(newRingTokenOwnershipInfo(token, initTokens[prev]))
+		prev = tk
+	}
+	heap.Init(tokensQueue)
+	tokensQueues[0] = tokensQueue
+
+	// instanceQueue is a priority queue of instances such that instances with higher ownership have a higher priority
+	instanceQueue := newPriorityQueue(t.instanceID)
+	instanceQueue.Add(newRingInstanceOwnershipInfo(0, instanceOwnership))
+	heap.Init(instanceQueue)
+	defer t.cleanUp(instanceQueue, tokensQueues)
+
+	for i := 1; i <= t.instanceID; i++ {
+		optimalInstanceOwnership := float64(1<<32) / float64(i+1)
+		currInstanceOwnership := 0.0
+		addedTokens := 0
+		tokens := make(Tokens, 0, tokensCount)
+		// currInstanceTokenQueue is the priority queue of tokens of newInstance
+		currInstanceTokenQueue := newPriorityQueue(tokensCount)
+		for addedTokens < tokensCount {
+			optimalTokenOwnership := t.getOptimalTokenOwnership(optimalInstanceOwnership, currInstanceOwnership, uint32(tokensCount-addedTokens))
+			highestOwnershipInstance := instanceQueue.Peek()
+			if highestOwnershipInstance.ownership < float64(optimalTokenOwnership) {
+				level.Error(t.logger).Log("msg", "it was impossible to add a token because the instance with the highest ownership cannot satisfy the request", "token", addedTokens+1, "highest ownership", highestOwnershipInstance.ownership, "requested ownership", optimalTokenOwnership)
+				break
+			}
+			tokensQueue := tokensQueues[highestOwnershipInstance.ringItem.GetID()]
+			highestOwnershipToken := tokensQueue.Peek()
+			if highestOwnershipToken.ownership < float64(optimalTokenOwnership) {
+				// token with the highest ownership of the instance with the highest ownership
+				// could not satisfy the request, hence we pass to the next instance.
+				heap.Pop(instanceQueue)
+				continue
+			}
+			token := highestOwnershipToken.ringItem.(*ringToken)
+			newToken, err := t.calculateNewToken(token, optimalTokenOwnership)
+			if err != nil {
+				return nil, err
+			}
+			tokens = append(tokens, newToken)
+			// add the new token to currInstanceTokenQueue
+			currInstanceTokenQueue.Add(newRingTokenOwnershipInfo(newToken, token.prevToken))
+
+			oldTokenOwnership := highestOwnershipToken.ownership
+			newTokenOwnership := float64(getTokenDistance(newToken, token.token))
+			currInstanceOwnership += oldTokenOwnership - newTokenOwnership
+			tokensQueue.Update(highestOwnershipToken, func(tokenOwnershipInfo *ownershipInfo) {
+				rT := tokenOwnershipInfo.ringItem.(*ringToken)
+				rT.prevToken = newToken
+				tokenOwnershipInfo.ownership = newTokenOwnership
+			})
+			instanceQueue.Update(highestOwnershipInstance, func(instanceOwnershipInfo *ownershipInfo) {
+				instanceOwnershipInfo.ownership = highestOwnershipInstance.ownership - oldTokenOwnership + newTokenOwnership
+			})
+			addedTokens++
+		}
+		if addedTokens != tokensCount {
+			return nil, errorNotAllTokenCreated(i, t.cfg.zone, tokensCount)
+		}
```

Comment:
As far as I understand, this can only happen if we perform the `break` above. So I'd move this there and remove the condition.